ush/python/pygw/src/pygw/executable.py

```diff
@@ -0,0 +1,354 @@
+import os
+import shlex
+import subprocess
+import sys
+from typing import Any, Optional
+
+__all__ = ["Executable", "which", "CommandNotFoundError"]
+
+
+class Executable:
+    """
+    Class representing a program that can be run on the command line.
+
+    Example:
+    --------
+
+    >>> from pygw.executable import Executable
+    >>> cmd = Executable('srun')  # Lets say we need to run command e.g. "srun"
+    >>> cmd.add_default_arg('my_exec.x')  # Lets say we need to run the executable "my_exec.x"
+    >>> cmd.add_default_arg('my_arg.yaml')  # Lets say we need to pass an argument to this executable e.g. "my_arg.yaml"
+    >>> cmd.add_default_env('OMP_NUM_THREADS', 4)  # Lets say we want to run w/ 4 threads in the environment
+    >>> cmd(output='stdout', error='stderr')  # Run the command and capture the stdout and stderr in files named similarly.
+
+    `cmd` line above will translate to:
+
+    $ export OMP_NUM_THREADS=4
```

Comment:
devil's advocate here, this is great but how can we convince Joe User that this is better than a bash script for this sort of thing?

Reply:
I hope that is convincing enough for provenance of the process itself.